lib/filterx/func-str.c

```diff
@@ -0,0 +1,473 @@
+        {
+          filterx_object_unref(needle_obj);
+          return NULL;
+        }
+      needles = g_ptr_array_sized_new(len);
+
+      for (guint64 i = 0; i < len; i++)
+        {
+          FilterXObject *elem = filterx_list_get_subscript(needle_obj, i);
+
+          GString *str_value = _obj_format(elem, self->ignore_case);
+          filterx_object_unref(elem);
+          if (!str_value)
+            goto error;
+          g_ptr_array_add(needles, str_value);
+        }
+      filterx_object_unref(needle_obj);
+      return needles;
+    }
+
+error:
+  g_ptr_array_free(needles, TRUE);
+  return NULL;
+}
+
+static FilterXObject *
+_expr_affix_eval(FilterXExpr *s)
+{
+  FilterXExprAffix *self = (FilterXExprAffix *)s;
+  FilterXObject *result = NULL;
+
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+  GString *haystack_str = _expr_format(self->haystack, self->ignore_case);
+  if (!haystack_str)
+    goto exit;
+
+  gboolean matches = FALSE;
+  const gchar *needle_str;
+  gsize needle_len;
+
+  if(self->needle.cached_strings->len != 0)
+    {
+      for(guint64 i = 0; i < self->needle.cached_strings->len && !matches; i++)
+        {
+          FilterXStringWithCache *current_needle = g_ptr_array_index(self->needle.cached_strings, i);
+          if (!_string_with_cache_get_string_value(current_needle, self->ignore_case, &needle_str, &needle_len))
+            goto exit;
+          matches = self->process(haystack_str->str, haystack_str->len, needle_str, needle_len);
+        }
+      result = filterx_boolean_new(matches);
+      goto exit;
+    }
+
+  GPtrArray *needle_list = _expr_affix_eval_needle(self);
+  if (!needle_list)
+    goto exit;
+
+  if(needle_list->len != 0)
+    {
+      for(guint64 i = 0; i < needle_list->len && !matches; i++)
+        {
+          GString *current_needle = g_ptr_array_index(needle_list, i);
+          matches = self->process(haystack_str->str, haystack_str->len, current_needle->str, current_needle->len);
+        }
+    }
+  g_ptr_array_free(needle_list, TRUE);
+  result = filterx_boolean_new(matches);
+
+exit:
+  scratch_buffers_reclaim_marked(marker);
+  return result;
+
+}
+
+static FilterXExpr *
+_extract_haystack_arg(FilterXFunctionArgs *args, GError **error, const gchar *function_usage)
+{
+  FilterXExpr *haystack = filterx_function_args_get_expr(args, 0);
+  if (!haystack)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "haystack must be set. %s", function_usage);
+      return NULL;
+    }
+  return haystack;
+}
+
+static FilterXExpr *
+_extract_needle_arg(FilterXFunctionArgs *args, GError **error, const gchar *function_usage)
+{
+  FilterXExpr *needle = filterx_function_args_get_expr(args, 1);
+  if (!needle)
+    {
+      g_set_error(error, FILTERX_FUNCTION_ERROR, FILTERX_FUNCTION_ERROR_CTOR_FAIL,
+                  "needle must be set. %s", function_usage);
+      return NULL;
+    }
+  return needle;
+}
```

Comment:
[nitpick]

These are guaranteed by the `FilterXFunctionArgs` class to succeed, because we checked the length before. I would recommend either removing the validations or changing them to assertions.

Reply:
Merged them into a single `_extract_args` method to better align with the codebase.